app/apps/handlers.js

```diff
@@ -19,18 +19,26 @@ exports.create = (req, res, next) => {
     repo_url: req.body.repo_url,
     userapps: [],
   });
-  let new_app;
+  const create_bucket = new Promise((resolve, reject) => {
+    if (req.body['new-app-datasource'] === 'create') {
+      resolve(new Bucket({
+        name: req.body['new-datasource-name']
+      }).create());
+    } else {
+      resolve(null);
+    }
+  });
 
-  app.create()
-    .then((created_app) => {
-      if(req.body['new-app-datasource'] === 'select') {
+  Promise.all([app.create(), create_bucket])
+    .then(([created_app, created_bucket]) => {
+      if (created_bucket) {
+        created_app.grant_bucket_access(created_bucket.id, 'readonly');
+      } else if(req.body['new-app-datasource'] === 'select') {
         created_app.grant_bucket_access(req.body['select-existing-datasource'], 'readonly');
```

Comment:
`grant_bucket_access` also returns a Promise, so you'll need to do something like:
```javascript
let grant_access = Promise.resolve(null);
if (created_bucket) {
  grant_access = created_app.grant_bucket_access(...);
} else if (...) {
  grant_access = created_app.grant_bucket_access(...);
}
grant_access.then(() => {
  res.redirect(url_for(...));
}
```